src/fd.cpp

```diff
@@ -0,0 +1,213 @@
+    }
+    return Timestamp(timeout);
+  }
+
+};
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+  const bool flag = args->flag->load();
+  args->flag->store(true);
+  if (flag)
+    return;
+  if (args->callback != nullptr) {
+    Rcpp::LogicalVector results(args->results.begin(), args->results.end());
+    (*args->callback)(results);
+  } else {
+    args->callback_native(args->results.data());
+  }
+
+}
+
+// CONSIDER: if necessary to add method for HANDLES on Windows. Would be different code to SOCKETs.
+// TODO: implement re-usable background thread.
+static int wait_thread(void *arg) {
+
+  tct_thrd_detach(tct_thrd_current());
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  // poll() whilst checking for cancellation at intervals
+
+  int ready;
+  double waitFor = std::fmax(args->timeout.diff_secs(Timestamp()), 0);
+  do {
+    waitFor = std::fmin(waitFor, 1.024); // set to just over the default 1s
+    ready = LATER_POLL_FUNC(args->fds.data(), args->num_fds, static_cast<int>(waitFor * 1000));
+    if (args->flag->load()) return 1;
+    if (ready) break;
+  } while ((waitFor = args->timeout.diff_secs(Timestamp())) > 0);
+
+  // store pollfd revents in args->results for use by callback
+
+  if (ready > 0) {
+    for (int i = 0; i < args->num_fds; i++) {
+      (args->results)[i] = (args->fds)[i].revents == 0 ? 0 : (args->fds)[i].revents & (POLLIN | POLLOUT) ? 1: NA_INTEGER;
+    }
+  } else if (ready < 0) {
+    std::fill(args->results.begin(), args->results.end(), NA_INTEGER);
+  }
+
+  callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
+
+  return 0;
+
+}
+
+static int execLater_launch_thread(std::shared_ptr<ThreadArgs> args) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(new std::shared_ptr<ThreadArgs>(args));
+
+  tct_thrd_t thr;
+
+  return tct_thrd_create(&thr, &wait_thread, static_cast<void *>(argsptr.release())) != tct_thrd_success;
+
+}
+
+static SEXP execLater_fd_impl(const Rcpp::Function& callback, int num_fds, struct pollfd *fds, double timeout, int loop_id) {
+
+  std::shared_ptr<ThreadArgs> args = std::make_shared<ThreadArgs>(callback, num_fds, fds, timeout, loop_id);
+
+  if (execLater_launch_thread(args))
+    Rcpp::stop("Thread creation failed");
+
+  Rcpp::XPtr<std::shared_ptr<std::atomic<bool>>> xptr(new std::shared_ptr<std::atomic<bool>>(args->flag), true);
+  return xptr;
+
+}
+
+// native version
+static int execLater_fd_impl(void (*func)(int *, void *), void *data, int num_fds, struct pollfd *fds, double timeout, int loop_id) {
+
+  std::shared_ptr<ThreadArgs> args = std::make_shared<ThreadArgs>(func, data, num_fds, fds, timeout, loop_id);
+
+  return execLater_launch_thread(args);
+
+}
```

Comment:
I think one of these should be renamed, it took me a little while to figure out how `execLater_fd` below returns `execLater_fd_impl`'s int return value and it becomes a cancellation token :)

Reply:
I've renamed the native version `execLater_fd_native` in [e0b21a9](https://github.com/r-lib/later/pull/190/commits/e0b21a949a556fac0ff5d1e87125a65892a85ef2).